desc/objectives/_omnigenity.py

```diff
@@ -501,6 +502,232 @@ def compute(self, params, constants=None):
+
+        assert grid_eq.NFP == grid_field.NFP
+        assert grid_eq.sym is False
+        assert grid_field.sym is False
+        assert grid_eq.num_rho == 1
+        assert grid_field.num_rho == 1
+
+        timer = Timer()
+        if verbose > 0:
+            print("Precomputing transforms")
+        timer.start("Precomputing transforms")
+
+        self._profiles = get_profiles(self._eq_data_keys, obj=eq, grid=grid_eq)
+        self._eq_transforms = get_transforms(
+            self._eq_data_keys,
+            obj=eq,
+            grid=grid_eq,
+            M_booz=M_booz,
+            N_booz=N_booz,
+        )
+        self._field_transforms = get_transforms(
+            self._field_data_keys,
+            obj=field,
+            grid=grid_field,
+        )
+        self._constants = {
+            "equil_transforms": self._eq_transforms,
+            "equil_profiles": self._profiles,
+            "field_transforms": self._field_transforms,
+            "helicity": self.helicity,
+        }
+
+        timer.stop("Precomputing transforms")
+        if verbose > 1:
+            timer.disp("Precomputing transforms")
+
+        if self._normalize:
+            # average |B| on axis
+            self._normalization = jnp.mean(field.B_lm[: field.M_well])
+
+        super().build(use_jit=use_jit, verbose=verbose)
+
+    def compute(self, equil_params, field_params, constants=None):
+        """Compute omnigenity errors.
+
+        Parameters
+        ----------
+        equil_params : dict
+            Dictionary of equilibrium degrees of freedom, eg Equilibrium.params_dict
+        field_params : dict
+            Dictionary of field degrees of freedom, eg OmnigenousField.params_dict
+        constants : dict
+            Dictionary of constant data, eg transforms, profiles etc. Defaults to
+            self.constants
+
+        Returns
+        -------
+        f : ndarray
+            Omnigenity error at each node (T).
+
+        """
+        if constants is None:
+            constants = self.constants
+        eq_data = compute_fun(
+            "desc.equilibrium.equilibrium.Equilibrium",
+            self._eq_data_keys,
+            params=equil_params,
+            transforms=constants["equil_transforms"],
+            profiles=constants["equil_profiles"],
+        )
+        field_data = compute_fun(
+            "desc.magnetic_fields.OmnigenousField",
+            self._field_data_keys,
+            params=field_params,
+            transforms=constants["field_transforms"],
+            profiles={},
+        )
+
+        # TODO: move to a compute function?
+        # would need to add iota as an OmnigenousField attribute
+        M, N = constants["helicity"]
+        iota = eq_data["iota"][0]  # FIXME: assumes a single flux surface
+        matrix = jnp.where(
+            M == 0,
+            jnp.array([N - M * iota, iota / N, 0, 1 / N]),
+            jnp.array([N, M * iota / (N - M * iota), M, M / (N - M * iota)]),
```

Comment:
do we also need to guard against division by zero on rational surfaces?

Reply:
Not in practice. Division by zero will only happen on the rational surface that corresponds to the field lines being parallel to the $B_{max}$ contour, for example a tokamak with no rotational transform.

Reply:
Wouldn't that be a pretty standard initial condition for QA/OT optimization?

Reply:
In that case don't we know what the boozer angles are? or we could just make some up. My point would be to just assign something non-degenerate, even if its a bit "wrong", just to get the optimizer started